src/settings/settinggroup.h

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef GROUPEDSETTINGS_H
+#define GROUPEDSETTINGS_H
+
+#include <QObject>
+#include <QVariant>
+
+#include "setting.h"
+
+/**
+ * @brief Represents a group of dynamic settings. Dynamic setttings can be
+ * defined at runtime.
+ *
+ * This abstraction serves as an aggregator of underlying storage settings.
+ * These settings necessarily need to be grouped under a QSettings group i.e.
+ * their keys need to share the same prefix followed by a slash ("/").
+ *
+ * Because settings under a SettingsGroup can be dynamically generated, upon
+ * initialization of a new SettingsGroup object the underlying storage is probed
+ * for any stored setting with the group prefix and those settings are
+ * registered to the SettingsManager.
+ *
+ * A SettingsGroups is useful when
+ *
+ * * Defining metrics dynamically at runtime is required or
+ * * Reacting to changes for _any_ setting under the group is necessary or
+ * * It makes sense to have metrics grouped for code semantics.
+ *
+ * It is allowed to create SettingGroups that are subsets or supersets of other
+ * SettingGroups. However, this can be tricky if the different groups are
+ * configured differently e.g. one has removeWhenReset set to true, while the
+ * other has removeWhenReset set to false. Whichever group register the
+ * underlying setting first will have their configurations enforced. There are
+ * Q_ASSERTs set up in debug mode to catch such situations.
+ *
+ */
+class SettingGroup : public QObject {
+  Q_OBJECT
+
+ public:
+  /**
+   * @brief Construct a new SettingGroup object
+   *
+   * @param groupKey The group key will be used as a prefix for the key of all
+   * settings added to the group e.g. groupKey/mySetting.
+   * @param removeWhenReset Whether or not this settings in this group should
+   * actually be removed when `reset` is called.
+   * @param sensitiveSetting Whether or not settings in this group are
+   * sensitive settings i.e. settings that must not be logged in plain text.
+   * @param acceptedKeys When this list has at least one member, only the listed
```

Comment:
Why is the enforcement using acceptedKeys needed? If not needed, should it and mayRecord() be removed?

Reply:
This is pre-work for experiments related stuff, which will require an allow list of keys.